preferences: rework

elstoc opened the pull request:
This post will be updated to reflect the discussions below / current state...

The preferences dialog has grown organically over time and is mostly two pages of miscellaneous preference options with lots of scrolling and hunting required to find what you're after.

I'd like to propose that we try to work out some way to bring order to the chaos.

I'm hoping that we can move many of the individual sections into their own tab and, to facilitate that I've started by moving the tabs to the left hand side of the dialog (to allow space for more tabs and make it easier to find a section quickly).

![Screenshot_2020-04-18_17-00-43](https://user-images.githubusercontent.com/9555491/79643060-fa259300-8198-11ea-8d32-97f90a89b08a.png)

To get the ball rolling there follows an initial suggestion for the new layout. The first level is the new tabs and the indented items below are the contents of those tabs - made up of sections or individual options from the existing tabs (their current location is in brackets).

For tabs that relate to darktable views (lighttable, darkroom, other views), the preferences dialog should open in the appropriate tab for the user's current view

```
interface (new tab)
  interface language (option: GUI/general)
  theme (option: GUI/general)
  modify theme with user tweaks (new option - await merge of #4670)
  css tweaks (new option to edit user.css - await merge of #4670)
import (new tab)
  import
    content of current section GUI/import
  session options 
    content of current tab: session options
lighttable (new tab)
  contents of current section: GUI/lighttable
  show scrollbars for center view (split from existing GUI/misc option)
  sort collection recent-older (option: GUI/misc)
  number of folder levels in lists (option: GUI/misc)
  overlay txt sidecar over zoomed images (option: GUI/misc)
  color manage cached thumbnails (option: core/quality)
darkroom (new tab)
  gui
    contents of current section GUI/darkroom
    show scrollbars for center view (split from existing GUI/misc option)
    demosaicing for zoomed-out (option: core/quality)
  core
    auto-apply basecurve & per-camera (options: core/quality)
    auto-apply sharpen (options: core/quality)
    pixel interpolator (option: core/quality)
    3D lut root folder (option: core/misc)
    use LittleCMS 2 to apply output color profile (option: core/quality)
other views (new tab)
  map/geolocalisation
    contents of current section: GUI/map/geolocalisation
  slideshow (new section)
    waiting time between each picture (option: GUI/misc)
    do high quality sampling (option: core/quality)
security (new tab)
  contents of current section: GUI/security
cpu/gpu/memory 
  contents of current section: core/cpu/gpu/memory
storage
  database
    various options: core/misc
  xmp
    xmp options (section: core/xmp)
miscellaneous (new tab)
  gui
    mouse wheel scrolls side panels by default (from GUI/misc)
    always show panels' scrollbars (from GUI/misc)
    method for getting display profile (option: GUI/misc)
  tags
    omit hierarchy in simple tag lists (option: core/misc)
    disable the entry completion (option: core/misc)
  other
    password storage backend (option: core/misc)
    audio executable (option: core/misc)
    april 1st game (option: gui/misc)
shortcuts (duplicate existing tab)
presets (duplicate existing tab)
```
I've almost certainly missed some things out, and there may be some 'hidden prefs' in darktablerc that might feel easier to incorporate in a less cluttered dialog.

discuss...

TurboGit commented:
I'm not fond of the vertical tabs. But I'm ok for adding more of them if this can help the situation, I do agree that the prefs is a bit crowded at this stage.

I would group the session and import group which are basically doing the same thing : doing something when new picture are added into collections.

- theme (including a text area to apply theme tweaks)
- lighttable (possibly combining some entries from gui and core)
- darkroom (similarly combining gui and core entries)
- session/import
- miscellaneous
- system (cpu/gpu/memory, database)
- shortcuts
- presets

There is also option to reorder, "overlay txt sidecar ovver zoomed images" which is i miscellaneous section where it could (should?) be in lighttable one.

We may also want to move the "method use for getting the display profile", currently in miscellaneous and would be better in core/quality. BTW, this option would be where in your new ordering? In lighttable ? In darkroom ? In system maybe ?

I think that we should first list all options that are in prefs now and decide where they should go. Probably some difficult decisions...

All this is also to get the ball rolling, I'm not clear at this stage.

elstoc commented:
Thanks @TurboGit. The reason I went for the vertical tabs was so we could fit more entries without running out of screen space or having it look too crowded, but perhaps there's a way to do this with multiple rows of horizontal tabs. Obviously it would need some styling - I'm certainly not a fan of the massive tabs in my screenshot - I'd want them grouped in the top-left of the screen and have the inactive ones in a different colour from the rest of the dialog.

The other advantage of the vertical tabs would be that we could present them in some semi-logical order (though what order I don't know).  I've certainly not fixed anything in my head at this stage.

johnny-bit commented:
Think not "vertical tabs", but settings menu :wink: 

![Zrzut ekranu z 2020-04-18 00-02-21](https://user-images.githubusercontent.com/10687765/79617482-e9790c80-8107-11ea-8af6-c155aa0a3724.png)

The way Gnome does it is "alright" I think.

Another pursuit for settings would be IMO to get rid of scrollbars for **most** option pages (obviously not prests/accels). That way cognitive load per setting "page" would be smaller, setting wouldn't "hide at the bottom" etc.

But yea - you've got the ball rolling @elstoc  :)

TurboGit commented:
@johnny-bit : right if the tabs can be made to look like standard GNOME pref dialog I'm ok.

johnny-bit commented:
I think it's a different control, but the idea's the same :)

sob., 18 kwi 2020 o 09:12 Pascal Obry <notifications@github.com> napisał(a):

> @johnny-bit <https://github.com/johnny-bit> : right if the tabs can be
> made to look like standard GNOME pref dialog I'm ok.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/darktable-org/darktable/pull/4747#issuecomment-615651839>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACRRKFMTZD3R3FYL37DS6LDRNFHE3ANCNFSM4MLAOBEQ>
> .
>


-- 
Pozdrawiam,
Hubert Kowalski

elstoc commented:
Yes I was hoping for something like the screenshot @johnny-bit posted. I'd hope to be able to style the existing notebook to look like that, but I can't work out how - css styling of gtk widgets is very new to me.  And I completely agree we need to try and avoid scrollbars where possible.

elstoc commented:
I've updated the OP with an initial suggestion for the new layout.

johnny-bit commented:
OK, I've found what I was looking for!

Try this: don't use notebook (it's for tabs), use Stack and for switching use StackSidebar.

check out gtk3-demo app with those and it fits the bill (at least on UI side ;)

hardcore ver would be to use treeview for switching ;)

elstoc commented:
Thanks @johnny-bit I'll have a play with that

phweyland commented:
My 2 cents.
Without any doubt a lot of configuration points have their place in centralized configuration as today and improvements as suggested above are welcome.
But some configurations could be closer to the related feature.
I think about map for example. I would find more convenient to have the configuration points available on map view, using a local configuration symbol (or not) like in metadata or export modules.
Another example is lut3d. I find boring to have to go to general configuration to set up the lut folder while it could be easily set up in the module.
What do you think ?

elstoc commented:
> use Stack and for switching use StackSidebar

Something like this (with the same 'needs CSS' caveats)...

![Screenshot_2020-04-18_17-00-43](https://user-images.githubusercontent.com/9555491/79642585-46bb9f00-8196-11ea-874d-8f1cd7257b15.png)

> But some configurations could be closer to the related feature.

Completely agree. Perhaps we should run some separate PRs alongside this one to move or duplicate those settings closer to the features (so that they can be removed in this change).

johnny-bit commented:
> Something like this (with the same 'needs CSS' caveats)...

Holly hell, yes :D 

now a bit splitting of settings and this is cool and cute :D

Nilvus commented:
> Yes I was hoping for something like the screenshot @johnny-bit posted. I'd hope to be able to style the existing notebook to look like that, but I can't work out how - css styling of gtk widgets is very new to me. And I completely agree we need to try and avoid scrollbars where possible.

If needed, just be sure to have Gtk part tweakable by CSS (so add identifiers as needed to tweak parts). For the CSS, I could work on that after.

And I agree to @johnny-bit and @TurboGit: if it's on Gnome like, it's ok, so at least having the left side a little darker than the main right side. By now, we don't see difference. And of course hover things will be needed.

TurboGit commented:
To follow on @phweyland idea of a context like preference, what about having a special tab that will contain the pref depending on the context, when pref opened from lighttable have there the lighttatble prefs, form map view the maps prefs... Using `ctrl+click `on the cranked wheel to open the pref we would open this tab by default ? For the filtering we may want to add another attribute in the XML dearktablerc description. Just thinking out loud :)

elstoc commented:
> when pref opened from lighttable have there the lighttatble prefs, form map view the maps prefs

If we have tabs with the same name as the views this is very simple and imo should be the default operation (no need to ctrl-click), since all of the tabs are named and you can select and activate them by name in the code.

houz commented:
I like this. Just one caveat: the original distinction between gui and core was to separate things that are only relevant when running dt with a gui vs. stuff that also affects dt-cli. so gui also had settings that were not "gui" settings but changed the behavior of dt.

TurboGit commented:
> If we have tabs with the same name as the views this is very simple and imo should be the default operation (no need to ctrl-click), since all of the tabs are named and you can select and activate them by name in the code.

But many prefs are not context related like OpenCL or database stuff. So we still want to get the complete pref dialog somehow.

elstoc commented:
Feels like this might bring us back to a long scrolling preferences dialog though, which I was trying to avoid.

The OpenCL (etc.) stuff will still be available as separate tabs it's just that the view-specific tab will open first.

houz commented:
No, splitting things is fine. I was just telling you why there was a differentiation. One could move things around, maybe we can come up with a way to mark settings that are not just related to the main darktable application. Just something to keep in mind.

johnny-bit commented:
I have 2 things that might be tried too...

1. I haven't seen option settings in any other app being repeated where they do matter in 2 categories. for example one setting is in "display" and another is in "video" and for me in that particular app creates confusion because in my system that settings are very related and changing one means i have to switch tab and change another... so it CAN be tried to have one setting in multiple places where it IS relevant... Just a thought tho ;)
2. Can we have simplest text-only (nothing fancy) display (or for fancier things) possibility to edit `darktablerc` directly in options in separate tab just for it (that tab existence could be another setting)

both ideas are just "thrown in the wind" don't pay attention to them ;)

houz commented:
I don't know what you mean with your point 1. I agree that settings should only be in one place, but I am not sure that's what you mean.
You can already edit darktablerc. Why would you want to add a text editor to dt to do that from within dt? It would also introduce potential conflicts between what you type in the text window and what you set in the gui version of the prefs. So for people who want to play with all the hidden config values I'd suggest using a text editor and reference `share/doc/darktable/darktablerc.html`.

elstoc commented:
> No, splitting things is fine...

Sorry I should have been clearer - I was responding to @TurboGit's comment:

> But many prefs are not context related like OpenCL or database stuff. So we still want to get the complete pref dialog somehow.

I do agree that OpenCL is a very common option that needs to be easily accessible, perhaps even with a shortcut to toggle it on/off. One option might be to have a 'sticky' panel at the bottom with OpenCL toggle on it? Though imo that's the only one I could think of that's important enough to always display, and if that's the case it might be better outside of the preferences dialog (bottom panel?).

> possibility to edit darktablerc directly in options

I'm not sure that would work at the moment, since (afaik) darktablerc is read once on opening darktable and written once on exiting (and its values are changed during use of darktable). I've always thought you had to be out of darktable for edits to darktablerc to have any effect. So perhaps we could do it but not call it 'editing darktablerc', rather 'editing a subset of internal config options' (or something).

elstoc commented:
BTW if you haven't seen I've put a suggestion for the new tab layout on the OP, so people have a starting point for discussion. Also I've pushed a commit incorporating the new darkroom/lighttable panels that auto-open based on the current view.

GLLM commented:
Hi all, @elstoc 

Just one idea, the display of the pattern for info line in DR, LT overlays & LT tooltip is way too small.
If the box could be higher (show 2-3 lines) or the pref window wider (show more in the line), it would be very nice !

Current display:
![image](https://user-images.githubusercontent.com/3989573/79688832-48966880-8251-11ea-9f94-f0d2eef201ed.png)

Here's my full customized line !
![image](https://user-images.githubusercontent.com/3989573/79688861-69f75480-8251-11ea-8805-1c68f33e95b2.png)


Thanks
GLLM

Nilvus commented:
@elstoc: I think this rework could also be the time to add a search field in preferences (I think especially about shortcuts and presets tabs actually).

Some parts of preferences could be merged together (the wanted CSS editor with actual theme menu and language one ; session options with import one actually in first tab). Some lines are also not at the best place, for example 'number of folders levels to display' could be on lighttable part, etc.)

elstoc commented:
> Some parts of preferences could be merged together (the wanted CSS editor with actual theme menu and language one ; session options with import one actually in first tab like @TurboGit already said!). Some lines are also not at the best place, for example 'number of folders levels to display' could be on lighttable part, etc.)

Agree - I've done some of this in my proposal above.

> add a search field

Um that might be a bit bigger a change than I intended. I'm not sure that would be possible without a complete rewrite. Perhaps not even then?

Nilvus commented:
> Um that might be a bit bigger a change than I intended. I'm not sure that would be possible without a complete rewrite. Perhaps not even then?

As you want. Search module could only be for shortcuts and presets settings. That's especially here this could the most useful. Other parts is less useful.

elstoc commented:
@Nilvus: shortcuts and presets are already searchable with ``ctrl+f``. However, for shortcuts, you need to first expand so that the thing you're searching for is visible.

Nilvus commented:
@elstoc: thanks, I didn't know this. Expanding first is not really convenient.

elstoc commented:
I agree. I get the feeling there might be a way to code the search without having to expand first but I can't work out how.

elstoc commented:
@Nilvus because it's an interactive search (it doesn't wait until you press enter) probably the best I can do is to have it automatically expand the first level of each tree (global, image operations, modules, views) when you start typing (after ctrl-f). This at least makes the search useful. What do you think?

Nilvus commented:
@elstoc: seems to be a good step.

elstoc commented:
> the display of the pattern for info line in DR, LT overlays & LT tooltip is way too small.
If the box could be higher (show 2-3 lines) or the pref window wider (show more in the line)

Should be resolved in latest commit. The preferences screen is now a little wider and the pattern entries should now span multiple lines if they don't fit.

elstoc commented:
I've done a bit of investigation and fleshed out the proposed layout in the OP and if nobody has any major objections I'm going to start reordering it tomorrow.

johnny-bit commented:
I don't mind nor do I object, however I do have one request if you may - please also post upgrades to OP in your posts - that way it's easier for the likes of me who checks mails more frequently than github and aren't fond of scrolling too much ;)

elstoc commented:
@johnny-bit ok will do

TurboGit commented:
Some comments:

>  xmp
>    xmp options (section: core/xmp)

Why not put that into the database section? Maybe renaming the section to **storage** ?

I would also keep the 
> shortcuts (duplicate existing tab)
> presets (duplicate existing tab)

The two last tabs.

elstoc commented:
@TurboGit thanks for your comments. I'm not sure that the xmp section sits nicely in the database since it's more to do with how the xmp _files_ are processed, though I did consider whether xmp and tag sessions might go nicely together in the same section.  Whatever we do, though, there will be some options that don't sit nicely in one section or another.

I'm happy to keep shortcuts/presets as the last two tabs.

On another note, having started work on it, the darkroom section is still quite long and I wonder whether splitting it into two sections "darkroom" and (something like) "processing" might be a better idea?

![Screenshot_2020-04-22_13-25-58](https://user-images.githubusercontent.com/9555491/79981593-d74af580-849c-11ea-8685-0b6d856b8e6d.png)

elstoc commented:
@TurboGit: sorry I missed the bit where you said rename to storage. That sounds fine.

TurboGit commented:
> On another note, having started work on it, the darkroom section is still quite long and I wonder whether splitting it into two tabs "darkroom" and (something like) "processing" might be a better idea?

At this stage it seems difficult to move things around but maybe we could have a more condensed theme for the preference dialog? There is lot of space between lines to me. I'm sure @Nilvus could help on this one if needed.

elstoc commented:
> There is lot of space between lines to me.

I agree. Also, I have increased the size of the dialog but we might still be able to make it bigger. What should we assume as a minimum screen size. Would it be safe to assume people are on at least 1080p monitors?

parafin commented:
> What should we assume as a minimum screen size. Would it be safe to assume people are on at least 1080p monitors?

No, people aren't. My MacBook Air has resolution of 1440×900.

elstoc commented:
> My MacBook Air has resolution of 1440×900

Ok I think I have it at 1200x800 at the moment. I won't make it any bigger.

elstoc commented:
Initial dev is now complete, though I'm sure there will still be further comments and changes. I've decided to split the darkroom tab in two - one for darkroom gui settings (the darktable tab) and one for settings that were previously in core (the processing tab). I've done this both for space reasons and because it keeps each tab having either core or gui options in it, but not both.

Inclusion of the CSS tweaks editor is waiting on #4670. 

For those who don't want to compile here's a bunch of screenshots of the new tabs...

![01_interface](https://user-images.githubusercontent.com/9555491/80027505-c7063b00-84db-11ea-99bb-cb7bea89b3e3.png)
![02_import](https://user-images.githubusercontent.com/9555491/80027516-cbcaef00-84db-11ea-8dc6-ffc78dc05b56.png)
![03_lighttable](https://user-images.githubusercontent.com/9555491/80027527-cec5df80-84db-11ea-8298-d332a94f8bd4.png)
![04_darkroom](https://user-images.githubusercontent.com/9555491/80027536-d1283980-84db-11ea-9817-7da32ba9a079.png)
![05_other_views](https://user-images.githubusercontent.com/9555491/80027542-d38a9380-84db-11ea-8676-6d3f152ddf6c.png)
![06_processing](https://user-images.githubusercontent.com/9555491/80027545-d5545700-84db-11ea-9e25-ff163378854c.png)
![07_security](https://user-images.githubusercontent.com/9555491/80027551-d7b6b100-84db-11ea-8b53-2522a1a0eddb.png)
![08_cpugpu](https://user-images.githubusercontent.com/9555491/80027557-dab1a180-84db-11ea-8b5a-4a9e14af1cdd.png)
![09_storage](https://user-images.githubusercontent.com/9555491/80027562-dd13fb80-84db-11ea-9984-7aed0002c3a2.png)
![10_misc](https://user-images.githubusercontent.com/9555491/80027576-e1401900-84db-11ea-8c61-70bd8842f65d.png)
![11_shortcuts](https://user-images.githubusercontent.com/9555491/80027587-e3a27300-84db-11ea-8a98-8589affd2671.png)
![12_presets](https://user-images.githubusercontent.com/9555491/80027593-e604cd00-84db-11ea-91f2-1fb4b00519df.png)

Nilvus commented:
@elstoc: no time to test that until this week-end at least. Before seeing that merged, did you add specific CSS tags names at least for left column and main one? It will be better to tweak them and have some visual background color difference to improve visibility of the window and separate parts.

TurboGit commented:
@Nilvus : I'll wait for your test on the week-end before merging.
@elstoc : This looks really nice ! I'll start looking at the code soon.

elstoc commented:
@Nilvus. Yes - everything is still underneath ``#preferences_notebook``, ``.sidebar`` is the entire sidebar and then you can use a ``:selected`` and ``:hover`` as descendant selectors to get the selected/hovered individual option. I mostly put those sections into darktable.css to document how to select them and assumed they'd basically be stubs that you would be able to fill out later.

As @TurboGit says it would be good to condense a little (get the prefs closer together) and it might be nice to get it so that the selected sidebar option always has the same background as the right-hand-side pane (and for the shortcuts/presets to have consistent coloring to the other panels). I don't think any of that necessarily needs to be in this PR though - you can do the theme stuff later.

@TurboGit thanks. I saw you merged #4670 now so I still have that user.css editor to add in. I'll look at that tomorrow.

elstoc commented:
@Nilvus for the remaining css selectors (including selecting the RHS panels), the following css:
```
#preferences_notebook stack
{
   border: 1px solid red;
}

#preferences_notebook 
{
   border: 1px solid blue;
}
```
Produces the following output:

![Screenshot_2020-04-22_22-13-03](https://user-images.githubusercontent.com/9555491/80034519-77793c80-84e6-11ea-8581-6ab82a1acfb4.png)

I'm not sure there's a way (currently) to select the entire dialog - it's not named - but I assume that's not an issue.

Nilvus commented:
@elstoc: I will check and test what I can done with that. Thanks for your screen capture and especially your last post. Will be easier with that. Anyway, I think that the sidebar should have a different grey that the main part on the right, like actual Gnome UI as shown by @johnny-bit in this PR.
And maybe I will be able to work on that a little sooner than expected. My work is calm actually, so more time for other things...

It's better that this PR is merged only after I made my first tests, just to be sure that I will not need something to tag as a CSS who could have been forgotten, in Gtk code. I don't have skills for that. But of course, my CSS work will be proposed as another PR (I have 2 other CSS issues to propose).

elstoc commented:
@Nilvus agreed.

Nilvus commented:
@elstoc: an issue on shortcuts search mode. It doesn't work on second or more child. For example, just test 'preview' shortcut. It doesn't find it but if you open the lighttable view then it works. Hope I'm clear here.

elstoc commented:
@Nilvus: The problem with the search mode the way it's currently implemented is that it only seems to be able to see visible nodes. I could make every node visible at the start but that would expand every child (including each individul module slider/combobox shortcut) and make the screen way too crowded (such that the only way to use it would be via the search). You also have to remember that the search is not a filter so it will find the first match and then stop - this means that if you search for "lighttable", for example, it will stop on the global shortcut and never be able to find its way further down the list.

My compromise is to restrict the search so that it only allows you to search for 'groups' of shortcuts at the moment. The only other way to do it would be to ditch the build-in search mechanism and build something from scratch that properly iterates through the children. It's possible but might be better as a separate PR and it might end up being worse for the reasons above. For now, at least it's better than what came before (which was utterly pointless).

Nilvus commented:
@elstoc: I understand and probably not easy to do. That's a good step but will need to be improved for the 3.2 release or we will have issues posted quickly.

Nilvus commented:
@elstoc: another thing on preferences window. **I would suggest to remove the dialog-action-box with the close button on preferences dialog window.** I check on some Gtk and Gnome apps, all preferences doesn't have close one. We close the window with the cross on the title bar. I think it's sufficient and will save some space on the window. That's actually make a whole line for just one button.

I would add that, as the @johnny-bit picture of Gnome preferences window (also without this button) that it will not be possible to really have same visual rendering with that title bar and having a different color on sidebar will be less good with that line.

elstoc commented:
Github doesn't seem to have caught up with my latest commit yet but when it does I've resolved the following

 - removed dialog action box & close button
 - moved the css ID ``#preferences_notebook`` so that it covers the entire dialog
 - added new section to interface tab to allow user.css to be edited and saved
 - changed the shortcut search label and tooltip text to make it clear that the search is only for 'shortcut groups' (e.g. module names and view names)

elstoc commented:
That did it. All up-to-date now.

Nilvus commented:
@elstoc: with Gtk Inspector tool, I still see a remaining part about a dialog-action box (less things than before your last commit) but still a part there. See:

![Capture-20200423172530-1518x911](https://user-images.githubusercontent.com/45535283/80117683-c3c88900-8587-11ea-9dfc-879090350e94.png)

elstoc commented:
As I understand it the action bar is part of the gtkdialog widget, though I assume it's possibly hidden/empty now.

I believe I can replace it with a hard-coded header bar (which replaces the one your window manager creates)

![Screenshot_2020-04-23_16-38-01](https://user-images.githubusercontent.com/9555491/80118839-d9868000-8580-11ea-994d-81d3fd695012.png)

Compared with

![Screenshot_2020-04-23_16-40-14](https://user-images.githubusercontent.com/9555491/80119074-210d0c00-8581-11ea-98bf-eece48ed6e9a.png)

Nilvus commented:
Another thing, the CSS edit view needs to be changed a little bit. Strangely, when I just change the theme (or language), I automatically move the pointer to click on save user CSS. That just disturbing.

This part needs at least two things:

- A better title than user.css above the textview. I would suggest "Your user personal CSS theme"
- Insert at least all CSS tweak part (from modify theme text and checkbox to "save user css" button) in a widget that allow to add borders or something like that. Things I could then tweak.

Nilvus commented:
> As I understand it the action bar is part of the gtkdialog widget, though I assume it's possibly hidden/empty now.
It's not the action bar (that needs to remains as it is) on what Gtk inspector tool show me. I have this selected when I click on the border bottom only with the pointer tool of Gtk Inspector tool. When doing that with this tool and hovering the preferences window, Gtk inspector show that I select all window except top action bar and clicking on the bottom of the window (where the close button was), the remaining tiny part is hovered by Gtk inspector tool and it select that part.

To be more clear, see the capture screen I made below (the remaining dialog-action-box no more useful is in yellow color):


Other thing, it will be better to not have the red margins on left, top and bottom of sidebar. But the margin are only because it's a box in a box. It will be better (as the Gnome UI shown above) to not have that. By removing the dialog-action-box no more useful, it will be easier to just have a sidebar and the right main part besides (without any widget/box margins that can be removed by CSS).

And with that, instead of applying a background color in whole preferences window then sidebar, it would be easier (and better) to set background color for sidebar then main right part.

So actually we have this tree:
![Capture-20200423182022-736x213](https://user-images.githubusercontent.com/45535283/80123409-283b1680-858f-11ea-8d0f-025002478711.png)

So instead of having this tree, it would be better to just have:

dialog
    stacksidebar
    stack

And so set background on stacksidebar and stack separately, without any shown margins around. Of course, if needed a box to put them inside (I don't know how Gtk works here) but without margins shown around stacksidebar and stack.

Hope it helps and it's clear.

elstoc commented:
> I would suggest "Your user personal CSS theme"

I want to avoid calling this a theme. The theme is whatever you select from the theme drop-down - this is just user-tweaks to that theme. Happy to rename but we need to think of something better.

elstoc commented:
> when I just change the theme (or language), I automatically move the pointer to click on save user CSS

I assume you mean that you automatically assume that the button means 'apply changes'? I suppose we could make it work like that.

> Insert at least all CSS tweak part (from modify theme text and checkbox to "save user css" button) in a widget that allow to add borders or something like that.

Sorry I'm not entirely sure what you're asking for here.

> So instead of having this tree, it would be better to just have...

I can't work out how to avoid the box (though I'm far from a gtk expert), but with some tweaks to the code plus some CSS I can get the borders to disappear. The following screenshot should show both the CSS tweaks plus the result (I've named the dialog container and the box that the stack/sidebar are packed into to make it easier).

![Screenshot_2020-04-23_19-09-25](https://user-images.githubusercontent.com/9555491/80134117-2fb1ee00-8596-11ea-8d77-2a915fca49ab.png)

Nilvus commented:
> > when I just change the theme (or language), I automatically move the pointer to click on save user CSS
> 
> I assume you mean that you automatically assume that the button means 'apply changes'? I suppose we could make it work like that.
> 
> > Insert at least all CSS tweak part (from modify theme text and checkbox to "save user css" button) in a widget that allow to add borders or something like that.
> 
> Sorry I'm not entirely sure what you're asking for here.

I'm not saying about how it works but how it's visually rendered that way. How it is visually by now is just disturbing as, by habit, it's easy to have the reflex to click on it to close the window (and apply changes), even if the only thing changed was one of the default themes or language (so without having made any tweaks).

So the problem is just some Gtk rework also here to add more options for CSS. As I said, I think the best thing would just integrate all CSS edit options you add in some Gtk widget to allow to differentiate visually that part (for example with borders I could tweak with CSS).

To see what I could do is something like that (don't care about the color, it's just to see part and border I would be able to add (and adjusting background color of the red selected part below):

![Capture-20200423210127-1120x753](https://user-images.githubusercontent.com/45535283/80138978-026d3c00-85a6-11ea-8fcb-a2c44d5ec112.png)

Nilvus commented:
> I can't work out how to avoid the box (though I'm far from a gtk expert), but with some tweaks to the code plus some CSS I can get the borders to disappear. The following screenshot should show both the CSS tweaks plus the result (I've named the dialog container and the box that the stack/sidebar are packed into to make it easier).
> 
> ![Screenshot_2020-04-23_19-09-25](https://user-images.githubusercontent.com/9555491/80134117-2fb1ee00-8596-11ea-8d77-2a915fca49ab.png)
For this part, just what I want. Perfect. Thanks.

Nilvus commented:
> > I would suggest "Your user personal CSS theme"
> 
> I want to avoid calling this a theme. The theme is whatever you select from the theme drop-down - this is just user-tweaks to that theme. Happy to rename but we need to think of something better.

Understand. So just call it "Your user personal CSS tweaks" (and having a specific tag like 'usercss' to allow specify bold font and/or different size also).

Said that, maybe it would be better to let the modify selected theme line with the checkbox with language and theme line so having only from the actual user.css: line to the save user css button on the widget/border I suggest, like that:

![Capture-20200423210127-1120x753](https://user-images.githubusercontent.com/45535283/80139720-3006b500-85a7-11ea-82a6-b64f35410c89.png)

That could add the possibility to only show (would be more consistent) to only show this edit part when the modify selected theme... checkbox checked.

So not check, all fields below not visible and if check these appears (the widget they're inside so.

elstoc commented:
> only show this edit part when the modify selected theme... checkbox checked

I did consider that. I do like being able to edit it without having to apply it first though, and being able to toggle it on and off without it disappearing.

elstoc commented:
I wonder if we need the "user.css" title at all. The default comment probably says enough for the user to know what it is. If we put the textview widget plus the save button in a box together (so they're visually separated from the rest of the dialog) I think that might be enough. What do you think?

Nilvus commented:
> > only show this edit part when the modify selected theme... checkbox checked
> 
> I did consider that. I do like being able to edit it without having to apply it first though, and being able to toggle it on and off without it disappearing.

That could also be improved later or on another PR or by something else if needed. Other things we discussed on last posts are, I think more necessary and needed. Waiting for that to work on CSS. And again, thanks for your work and time for that part.

Nilvus commented:
> I wonder if we need the "user.css" title at all. The default comment probably says enough for the user to know what it is. If we put the textview widget plus the save button in a box together (so they're visually separated from the rest of the dialog) I think that might be enough. What do you think?

Also, indeed that text is not needed. So, if you remove it I would suggest that:

Integrate the modify selected theme with checkbox inside the border I would like to have (so my first mockup). Then, that needs to change the modify line text. I suggest:

**modify selected theme with tweaks added below**

elstoc commented:
Ok so here it is with CSS so you can see how its made up:

![Screenshot_2020-04-23_20-49-12](https://user-images.githubusercontent.com/9555491/80143033-144ddf80-85a4-11ea-8acf-2f6bc8e3e1f7.png)

and with the default text:

![Screenshot_2020-04-23_20-49-57](https://user-images.githubusercontent.com/9555491/80143075-2596ec00-85a4-11ea-9cc9-9eb3b2727a41.png)

Good for you?

elstoc commented:
Or perhaps I misunderstood and you wanted that as the text for the checkbox?

elstoc commented:
This is better (?):

![Screenshot_2020-04-23_21-07-12](https://user-images.githubusercontent.com/9555491/80144469-760f4900-85a6-11ea-8f3b-a3e6bbd13a96.png)

elstoc commented:
I've pushed it anyway. Can always change again later!

Nilvus commented:
> Or perhaps I misunderstood and you wanted that as the text for the checkbox?

Yes you misunderstood and that was indeed the text for checkbox. So much better and perfect on second try (last screenshot). Thanks, i will check and test what I can do with CSS and your last commit.

elstoc commented:
> @elstoc: an issue on shortcuts search mode. It doesn't work on second or more child. For example, just test 'preview' shortcut. It doesn't find it but if you open the lighttable view then it works. Hope I'm clear here.

@Nilvus try the latest commit - I worked out a way to do a proper search

Nilvus commented:
> > @elstoc: an issue on shortcuts search mode. It doesn't work on second or more child. For example, just test 'preview' shortcut. It doesn't find it but if you open the lighttable view then it works. Hope I'm clear here.
> 
> @Nilvus try the latest commit - I worked out a way to do a proper search (and now I've also fixed it so it builds!)

Good work, it's great now. Apply on all levels. I would just suggest to add a tooltip on search button with something like: "click again to go to next result if available"

I've try with preview keyword and clicking again let go to the next place where preview keyword is. Users will not always think of clicking again.

Nilvus commented:
@elstoc (and others): with last changes you made, I've made a first draft of the window and especially CSS edit view. I've made few tries for that and I think the best one is this: integrate visually by playing with background-color and border to see the save user css button on the textview. Of course, it's just visually and I think that made all this part as a one thing. This way, if something add a lot of tweaks, the scrollbar will appear and see the separation.

What do you think of that? See below with grey theme and dark theme:

![Capture-20200424175556-1094x746](https://user-images.githubusercontent.com/45535283/80232288-fa200a00-8654-11ea-8475-1b54f5d8e55a.png)

![Capture-20200424175619-1094x746](https://user-images.githubusercontent.com/45535283/80232301-fb513700-8654-11ea-9810-b26c07b5de70.png)

There's still some margins and other things to work on.

Nilvus commented:
@elstoc: I've just had two times darktable crash when closing the window with the crash.

First time on Terminal, the error message was simply, lines before were not related:

```
double free or corruption (out)
Abandon
```

And second time, with more things:
```
#57 0x00007feec01a6533 in g_main_loop_run () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#58 0x00007feebfbf330d in gtk_main () at /lib/x86_64-linux-gnu/libgtk-3.so.0
#59 0x00007feec087112d in dt_gui_gtk_run (gui=<optimized out>) at /home/nilvus/Projets/dtprefs/src/gui/gtk.c:1454
        widget = <optimized out>
        allocation = {x = 0, y = 0, width = 568, height = 468}
#60 0x0000563df0aac0d6 in main (argc=<optimized out>, argv=<optimized out>) at /home/nilvus/Projets/dtprefs/src/main.c:93
[Inferior 1 (process 761353) detached]
backtrace written to /tmp/darktable_bt_L8ZMJ0.txt
Erreur de segmentation
```
And the text file: 
[darktable_bt_L8ZMJ0.txt](https://github.com/darktable-org/darktable/files/4530071/darktable_bt_L8ZMJ0.txt)

elstoc commented:
> "click again to go to next result if available"

Good idea. FYI you don't even need to use the button really - you can just press the enter key.

CSS edit view looks good. Would be even better if the currently selected option in the sidebar was the same colour as the background of the stack.

Nilvus commented:
> > "click again to go to next result if available"
> 
> Good idea. FYI you don't even need to use the button really - you can just press the enter key.

So this info is to add on the tooltip: "click or press enter again..."

> CSS edit view looks good. Would be even better if the currently selected option in the sidebar was the same colour as the background of the stack.

It's just a first draft, lot of things like that to work on. I don't spend time actually on that part. Margin on text in the interface view need to be improve (and other things...)

elstoc commented:
> double free or corruption (out)
   Abandon

Yeh I might need someone to give me some assistance on that. I think it's related to the new struct I've created but I'm not sure of the best way to free the memory - @TurboGit? Still new to some of this.

elstoc commented:
@Nilvus try the last commit - generally needs a code review by someone more knowledgeable than me but I think I've resolved the errors I can reproduce.

TurboGit commented:
@elstoc : I'll look at the memory during the week-end. I'll let you know.

elstoc commented:
@TurboGit thanks

TurboGit commented:
I tried to reproduce and was not able. @Nilvus : anything special done to get the double free ?

Also I think that on the GUI part we should have more space at the start of the buttons and in the before/after column names, a little screenshot to help:

![Capture d’écran de 2020-04-24 19-46-03](https://user-images.githubusercontent.com/467069/80241969-d82e8380-8664-11ea-9980-bb087c198bb5.png)

TurboGit commented on src/gui/preferences.c:
```diff
@@ -806,6 +938,74 @@ static void update_accels_model(gpointer widget, gpointer data)
+
+  gtk_tree_view_collapse_all(GTK_TREE_VIEW(tv));
+
+  int siblings = gtk_tree_model_iter_n_children(GTK_TREE_MODEL(tvmodel), NULL);
```
const

elstoc commented:
I couldn't either but there were a couple of other errors that I fixed when I force-re-pushed.

I think my main question is that I have a gchar pointer in my struct but I'm potentially replacing its multiple times with new strings and I don't quite know how that works with regard to the previous data.

elstoc commented:
Thanks @TurboGit. All done.

Nilvus commented:
> I tried to reproduce and was not able. @Nilvus : anything special done to get the double free ?

That's the difficulty. I've had the crash only two times and I've closed the preferences dialog window probably 5 or 6 times this day. So not really reproducible.
 
> Also I think that on the GUI part we should have more space at the start of the buttons and in the before/after column names, a little screenshot to help:

That's on my todo-list, as other things.

elstoc commented:
> I've had the crash only two times and I've closed the preferences dialog window probably 5 or 6 times this day

Have you tried the latest commit? It's fixed a few things that might be related. If I recall correctly I did have that error in an earlier version because I was trying to free an uninitialised pointer and I _think_ I've resolved most of those sorts of things now.

Nilvus commented:
No, I just see your commit a few minutes ago. I'm on another CSS issue actually.

But just to say that I've had a third crash (without last commit) by closing preferences and the error on terminal was just:

```
free(): invalid pointer
Abandon
```
Quite strange that the errors are not always the same...

elstoc commented:
I've seen that error and I'm pretty sure that's fixed now. Anyway update when you get a chance - I've not had any issues since.

elstoc commented:
One more minor tweak to make the spacing more consistent. The checkboxes still seem slightly off but I'm not certain how to resolve that.

Nilvus commented:
> One more minor tweak to make the spacing more consistent. The checkboxes still seem slightly off but I'm not certain how to resolve that.

WIth css probably. Be patient, it's a work in progress.

A new step was done yesterday for something you asked:
![Capture-20200425114104-244x699](https://user-images.githubusercontent.com/45535283/80276527-c216d680-86e9-11ea-9091-b8090dd4165b.png)

elstoc commented:
Looking good. 

It may be my misunderstanding but gtk seems to hard-code default spacing/borders/margins that can't be adjusted with CSS - or at least that was what I found when looking at the previous 'box' issues.  The spacing between preference items seems to be hard-coded in a similar way - 3px at the moment (reduced from 5) so you may hit those limits.

TurboGit commented:
> A new step was done yesterday for something you asked:

Nice !

Nilvus commented:
> It may be my misunderstanding but gtk seems to hard-code default spacing/borders/margins that can't be adjusted with CSS - or at least that was what I found when looking at the previous 'box' issues. The spacing between preference items seems to be hard-coded in a similar way - 3px at the moment (reduced from 5) so you may hit those limits.

I don't think so. Gtk Inspector Tool remains the best tool to check that but don't forgot also that lot of margins are also already sets as margins exist for textview, buttons, checkboxes. Anyway, if I see something I can't change by CSS, be sure I will come back to you here. Stay tuned!

Nilvus commented:
@elstoc: just a thing about some name. First category is called interface and on miscellaneous one you set "gui" sub-category. Interface and gui are quite same name, so I think "gui" should be renamed "display" to make a difference and still be related to the related options.

elstoc commented:
Problem is that the word 'display' doesn't really make sense in context of the "mouse wheel scrolls side panels by default" option, but 'gui' (or indeed, 'interface') does.  We could name that first tab back to 'general' (which is what the language/theme section was named before) and then rename 'gui' to 'interface'?

elstoc commented:
or we could rename the first section 'language/theme'

Nilvus commented:
> Problem is that the word 'display' doesn't really make sense in context of the "mouse wheel scrolls side panels by default" option, but 'gui' (or indeed, 'interface') does. We could name that first tab back to 'general' (which is what the language/theme section was named before) and then rename 'gui' to 'interface'?

That's also a possibility.

Another thing, how you render the 3 first options on first category (language, theme and modify... lines) are not made same way than on other categories. That's make CSS tweaks not consistent between them. 
First and more important one is that all labels are put in a widget, except the 2 lines theme and "modify selected theme". Could you add also those 2 labels in a widget?
This will allow to have same margin between text and option beside (checkbox, dropdown menu...) for all options on preferences window at same time.

Second thing I see (but maybe it's not possible to change that) is that you put all items in that same category into a box but all others categories are in a scrolledwindow. Same, adding that on scrolledwindow will allow to use same margin for all categories and avoid forgot tweak as actually it needs to set differently those. So if it is possible, it would be better and allow to also use only one thing for settings all same things (here same margins with one setting for all categories content).

Nilvus commented:
@elstoc: you could forgot second point if too difficult or problematic to change from box to scrolledwindow for first category. I finally made it a one set with CSS.

Most important thing remains to insert labels in a widget.

elstoc commented:
> you could forgot second point

Good - that was going to be hard to do

> First and more important one is that all labels are put in a widget

I think I've done what you wanted but it's hard because the first tab is hard-coded and in a slightly different way to the rest of the tabs. I don't think it's a really big deal if it looks a little different.

As discussed I've renamed the first tab to 'general' and misc/gui to misc/interface

Nilvus commented:
> > First and more important one is that all labels are put in a widget
> 
> I think I've done what you wanted but it's hard because the first tab is hard-coded and in a slightly different way to the rest of the tabs. I don't think it's a really big deal if it looks a little different.
> 
> As discussed I've renamed the first tab to 'general' and misc/gui to misc/interface

Should be good, thanks. I think it's a big deal. I've done a lot of work these past months (to complete big work by @aurelienpierre) to have a consistent UI and clean UI. And this is all about that.

Nilvus commented:
@elstoc: I'm almost done but 2 things remains:

- could you add a css specific tag on both box widgets where are the buttons at the bottom of presets and shortcuts. I can't set these box buttons (even with the parent > css selector) without breaking something else.

- I don't understand why there's so much differences on some categories from where options are placed. Is there's still hardcoded width? That's not consistent. I've check and all have same margins and paddings (on what I see and test, I don't think it's css related).

See below:

In general, as you can see drop-down menus not use the whole remaining length space
![general](https://user-images.githubusercontent.com/45535283/80287095-c534b580-872f-11ea-82c5-e6a4f14a2f25.png)

On processing view, you could see that drop-down menus use the whole remaining length space:
![processing](https://user-images.githubusercontent.com/45535283/80287130-fd3bf880-872f-11ea-9d69-bb2a52e8c157.png)

I don't really know which one of these is the better (whole remaining space or not)...

On CPU/GPU/Memory it's same problem but with textview, they're not using same length. If you check on other views, textview use whole remaining space.
![cpugpumemory](https://user-images.githubusercontent.com/45535283/80287208-7dfaf480-8730-11ea-81b7-b37f685b8da9.png)

Last thing, there's a bigger margin/padding between label widget and options besides (checkbox, textview, etc.) that's not there on other views. See below:
![storage](https://user-images.githubusercontent.com/45535283/80287252-d6ca8d00-8730-11ea-9492-826d4c0fe296.png)

![otherviews](https://user-images.githubusercontent.com/45535283/80287257-df22c800-8730-11ea-8e2d-87cb08218852.png)

Nilvus commented:
So saying that, 2 things needs to be decided:

- drop-down menus should use whole remaining length space or not?
- Is spinbutton entries who only have number settings should use whole remaining length space or not?

Nilvus commented:
And @TurboGit: is this new header for presets and shortcuts (here presets view) is ok for you:

![Capture-20200425202056-1099x243](https://user-images.githubusercontent.com/45535283/80287504-66247000-8732-11ea-958a-536c4618d7dd.png)

TurboGit commented:
Looks perfect! Thanks.

elstoc commented:
> In general, as you can see drop-down menus not use the whole remaining length space

The comboboxes in general are built differently to how the others are because its content is entirely hard-coded, whereas the other tabs (except shortcuts/presets) are auto-built. I don't entirely understand how the auto-built ones work (I didn't have to change them for this PR so I didn't look at them in any detail.

> there's a bigger margin/padding between label widget and options besides (checkbox, textview, etc.) that's not there on other views

I think that might be because they have labels on them, that need to sit in a box, and then the subsequent widgets/labels also sit in the same box. I can make the others consistent so they're all sat in a box as well (so at least you could style them in a consistent way).

elstoc commented:
> I think that might be because they have labels on them, that need to sit in a box...

Scratch that. I don't know

Nilvus commented:
> > I think that might be because they have labels on them, that need to sit in a box...
> 
> Scratch that. I don't know

Hmm, on what I see with Gtk Inspector too, the structure was the same (or maybe I miss something).

elstoc commented:
So in conclusion...

> could you add a css specific tag on both box widgets where are the buttons at the bottom of presets and shortcuts

You mean the box that surrounds the buttons rather than the buttons themselves? No problem - I'll name them ``#shortcut_controls`` and ``#preset_controls``.

> I don't understand why there's so much differences on some categories from where options are placed.

Apart from the caveat that the drop-downs in the 'general' tab are constructed differently I also have no idea. The code that constructs the widgets/labels on each of the 'non-general' tabs is identical. I've tried changing some things but I've hit a dead-end and nothing has worked.

Also I think the only things that should really span the whole space are strings and longstrings. But I don't know how to do that given the inconsistencies between unless you can find a pure-css solution.

elstoc commented:
I finally found the issue. Each of the preferences tabs is split into a two-column grid. The size of each of the data entry widgets was set to the same size as the grid column, which depends on what other widgets are present. So if there was a text preference present, this caused all of the widgets to expand to fill the screen width.

The solution was to pack all of the widgets into a box and choose which ones to expand-to-fill and which not to.

Latest commit should be consistent in all tabs.

elstoc commented:
Something like the following seems to make it all seem a little less ragged when you go into it...

```
#preferences_notebook combobox,
#preferences_notebook spinbutton
{
min-width: 150px;
}
```

Nilvus commented:
Thanks @elstoc. I've just compiled your last commit. Much better. I will try your CSS tip. I should be all done in the next hour. And if all ok, I think this PR could be then merged. Great work again, these new prefs are much better than old one.

Nilvus commented:
@elstoc and @TurboGit: this all ok for me. I think this PR could be merged on master. CSS work is almost done: just some minor cleaning to do, all things are set. As soon as this PR is merged, I could propose my CSS PR.

TurboGit commented:
Ok to me. I'm now merging that. @Nilvus : let's go for the CSS PR :)

elstoc commented:
FYI I've had an email from a user who would like to remove the GTK_DIALOG_USE_HEADER_BAR setting (which adds the hard-coded header bar with the 'X') and I have some sympathy with this position - the header bar takes up a lot of space and some people prefer to reduce or entirely remove window decorations, and also the ESC key gives the user an alternative way out of the window. This option is generally designed to allow you to include other buttons in that bar and we are not doing that here.

I removed it because the 'action area' was still present on the bottom of the preferences window and I agree it might be better to revert back, assuming we can keep that action area invisible. Once the CSS changes are merged I'll do some experimentation and possibly change it back as part of #4424, for which I already need to change the prefs screen.

TurboGit commented:
Well est `ESC` key should be to leave without changing the prefs in fact. Probably a nice new feature if we can do that. Maybe by storing all settings before entering the dialog and if exited with `ESC` restore all values. But in no was `ESC` should be used to exit & validate the changes.

elstoc commented:
I think ``ESC`` or the close button previously did exit without saving but removal of the close button meant that had to be changed. There is now no way to exit without changing because I can only attach the preferences response callback to ``GTK_RESPONSE_DELETE_EVENT``. The close button was associated with ``GTK_RESPONSE_ACCEPT`` which committed the changes.

Nilvus commented:
@elstoc: I think remove the header bar is not a good idea. I know users who prefer using mouse instead of the keyboard. They will not like seeing the header bar removed. I was about posting my CSS PR but I think I can do something to reduce space this headerbar take. I think this will be better.

Said that, having the possibility to close the window also with `ESC` is a good idea.

ptilopteri commented:
@Nilvus 
> Said that, having the possibility to close the window also with `ESC` is a good idea.

but then one would need a method to determine if <esc> saved or discarded changes :(

elstoc commented:
The only way to achieve the former behaviour without reinstating the big close button at the bottom of the window would be to add an 'apply' button to the header bar. Then ``ESC`` and ``X`` could have their former behaviours of closing without saving. 

> I think remove the header bar is not a good idea.

This wouldn't remove the header bar, just remove the **hard-coded gtk header bar** and replace it with the one that the user's window manager generates.

elstoc commented:
> The only way to achieve the former behaviour

Like this

![Screenshot_2020-04-26_14-46-27](https://user-images.githubusercontent.com/9555491/80309410-ca940d80-87cc-11ea-9f94-e7bb9c52265f.png)

johnny-bit commented:
change "apply" to "save" and we're good ;)

elstoc commented:
Ok I'll submit another PR

Nilvus commented:
@elstoc: excellent that way. About reducing headerbar, I can't remove more than 2px the height, and by adding few lines. So I will not integrate that. I don't know if a minimal height is hard-coded or not. Anyway with your new save button, that's better.

elstoc commented:
See #4821
